File: src/services/posService.js
/**
 * ============================================================
 * FIL: src/services/posService.js
 * FORMÅL: Simulerer sending av en ordre til PC-kassen (POS-systemet).
 *
 * HVA ER EN "SERVICE"?
 *  I backend-arkitektur er en "service" en klasse eller modul
 *  som håndterer én type forretningslogikk.
 *  Denne servicen sin jobb er å:
 *   1. Ta imot en ferdig mappet POS-ordre
 *   2. Bygge SOAP XML av den
 *   3. "Sende" den til POS-systemet (her: bare skrive det ut)
 *
 * I VIRKELIGHETEN ville sendTilPos() gjort et HTTP POST-kall
 * til POS-systemets SOAP-endepunkt med XML-en som body.
 *
 * NYT I DEL 2:
 *  - skrivUtOrdre() viser nå alle fraktlinjer og rabatter
 * ============================================================
 */

const { byggSoapXml } = require("../utils/xmlBuilder");

/**
 * skrivUtOrdre
 * ------------
 * Skriver POS-ordren ut i et lesbart format i konsollen.
 * Oppdatert i del 2 til å vise fraktLinjer og rabatter.
 *
 * @param {object} posOrdre - Mappet POS-ordre
 */
function skrivUtOrdre(posOrdre) {
  console.log("\n=== POS-ORDRE (lesbart format) ===");
  console.log(`Ordre-ID:   ${posOrdre.ordreId}`);
  console.log(`Kunde:      ${posOrdre.kundenavn}`);
  console.log(`E-post:     ${posOrdre.kundeEpost}`);

  // DEL 2: Vis alle fraktlinjer (ikke bare én)
  console.log("\nFraktlinjer:");
  posOrdre.fraktLinjer.forEach((linje, index) => {
    const prisVisning = linje.pris === 0 ? "gratis" : `kr ${linje.pris},-`;
    console.log(`  ${index + 1}. ${linje.metode} (${prisVisning})`);
  });
  console.log(`  Total frakt: kr ${posOrdre.totalFrakt},-`);

  // DEL 2: Vis rabatter hvis de finnes
  if (posOrdre.rabatter.length > 0) {
    console.log("\nRabatter:");
    posOrdre.rabatter.forEach((rabatt) => {
      console.log(`  - Kode: ${rabatt.kode} → -kr ${rabatt.belop},- (${rabatt.type})`);
    });
    console.log(`  Total rabatt: -kr ${posOrdre.totalRabatt},-`);
  }

  console.log(`\nTOTAL:      kr ${posOrdre.totalBelop},-`);

  console.log("\nProduktlinjer:");
  posOrdre.produkter.forEach((produkt, index) => {
    console.log(
      `  ${index + 1}. ${produkt.produktnavn} [${produkt.sku}]` +
        ` – ${produkt.antall} stk x kr ${produkt.enhetspris},- = kr ${produkt.linjetotal},-`
    );
  });
}

/**
 * sendTilPos
 * ----------
 * Simulerer sending av ordren til PC-kassen via SOAP.
 * I en ekte integrasjon ville dette vært et HTTP-kall med XML som body.
 *
 * @param {object} posOrdre - Mappet POS-ordre
 * @returns {boolean} - true hvis sending "lykkes"
 */
function sendTilPos(posOrdre) {
  console.log("\n=== SENDER TIL POS-SYSTEM ===");
  console.log("Bygger SOAP XML...");

  const soapXml = byggSoapXml(posOrdre);

  console.log("Kobler til POS-endepunkt: http://pos-system.intern/soap/ordre");
  console.log("Sender melding...\n");

  console.log("=== SOAP XML (dette ville blitt sendt) ===");
  console.log(soapXml);

  console.log("\n✓ POS-systemet svarte: 200 OK – Ordre mottatt og registrert");

  return true;
}

// ---------------------------------------------------------------
// EKSPORT
// ---------------------------------------------------------------
module.exports = { sendTilPos, skrivUtOrdre };

/**
 * ============================================================
 * OPPSUMMERING:
 *  - skrivUtOrdre(): printer ordren lesbart inkl. alle fraktlinjer og rabatter
 *  - sendTilPos(): simulerer sending av SOAP-melding til PC-kassen
 *  - I en ekte integrasjon ville sendTilPos() gjort HTTP POST til SOAP-URL
 * ============================================================
 */

File: dotnet/ShopifyPosIntegrasjon/Services/PosService.cs
// ============================================================
// FIL: Services/PosService.cs
// FORMÅL: Viser og "sender" en ordre til POS-systemet.
//         Tilsvarer posService.js i Node.js-versjonen.
//
// SAMMENLIGNING MED JAVASCRIPT:
//  JS:  posOrdre.fraktLinjer.forEach((linje, index) => { ... })
//  C#:  foreach (var linje in posOrdre.FraktLinjer) { ... }
//       (index: bruk en teller-variabel, eller Select med indeks)
//
//  JS:  console.log(`${a}: ${b}`)
//  C#:  Console.WriteLine($"{a}: {b}")
// ============================================================

namespace ShopifyPosIntegrasjon.Services;

public class PosService
{
    // XmlBuilder injiseres som en avhengighet (dependency)
    // I fremtidige prosjekter vil dette gjøres via DI-container
    private readonly XmlBuilder _xmlBuilder = new();

    // Tilsvarer "function skrivUtOrdre(posOrdre)" i JS
    public void SkrivUtOrdre(PosOrder posOrdre)
    {
        Console.WriteLine("\n=== POS-ORDRE (lesbart format) ===");
        Console.WriteLine($"Ordre-ID:   {posOrdre.OrdreId}");
        Console.WriteLine($"Kunde:      {posOrdre.Kundenavn}");
        Console.WriteLine($"E-post:     {posOrdre.KundeEpost}");

        // Vis alle fraktlinjer
        Console.WriteLine("\nFraktlinjer:");
        for (int i = 0; i < posOrdre.FraktLinjer.Count; i++)
        {
            var linje = posOrdre.FraktLinjer[i];
            // Ternary operator: condition ? verdi_hvis_true : verdi_hvis_false
            // Tilsvarer linje.pris === 0 ? "gratis" : `kr ${linje.pris},-` i JS
            string prisVisning = linje.Pris == 0 ? "gratis" : $"kr {linje.Pris},-";
            Console.WriteLine($"  {i + 1}. {linje.Metode} ({prisVisning})");
        }
        Console.WriteLine($"  Total frakt: kr {posOrdre.TotalFrakt},-");

        // Vis rabatter hvis de finnes
        if (posOrdre.Rabatter.Count > 0)
        {
            Console.WriteLine("\nRabatter:");
            foreach (var rabatt in posOrdre.Rabatter)
                Console.WriteLine($"  - Kode: {rabatt.Kode} → -kr {rabatt.Belop},- ({rabatt.Type})");
            Console.WriteLine($"  Total rabatt: -kr {posOrdre.TotalRabatt},-");
        }

        Console.WriteLine($"\nTOTAL:      kr {posOrdre.TotalBelop},-");

        Console.WriteLine("\nProduktlinjer:");
        for (int i = 0; i < posOrdre.Produkter.Count; i++)
        {
            var p = posOrdre.Produkter[i];
            Console.WriteLine(
                $"  {i + 1}. {p.Produktnavn} [{p.Sku}]" +
                $" – {p.Antall} stk x kr {p.Enhetspris},- = kr {p.Linjetotal},-"
            );
        }
    }

    // Tilsvarer "function sendTilPos(posOrdre)" i JS
    // bool = boolsk verdi (true/false), tilsvarer boolean i JS
    public bool SendTilPos(PosOrder posOrdre)
    {
        Console.WriteLine("\n=== SENDER TIL POS-SYSTEM ===");
        Console.WriteLine("Bygger SOAP XML...");

        string soapXml = _xmlBuilder.ByggSoapXml(posOrdre);

        Console.WriteLine("Kobler til POS-endepunkt: http://pos-system.intern/soap/ordre");
        Console.WriteLine("Sender melding...\n");

        // I virkeligheten: await httpClient.PostAsync(url, new StringContent(soapXml))
        Console.WriteLine("=== SOAP XML (dette ville blitt sendt) ===");
        Console.WriteLine(soapXml);

        Console.WriteLine("\n✓ POS-systemet svarte: 200 OK – Ordre mottatt og registrert");

        return true;
    }
}
